### PR TITLE
[CUDA] Support attention_bias in GroupQueryAttention via the unfused path

### LINKED ISSUE #29506: [Feature Request] CUDA EP: support `attention_bias` in GroupQueryAttention (last EP missing it)
### Describe the feature request

`com.microsoft.GroupQueryAttention` accepts an optional `attention_bias` input (input #10, "additional add to QxK'"). Today every actively-used EP implements it **except CUDA**, which rejects it at runtime:

```
Non-zero status code returned while running GroupQueryAttention node.
Name:'/model/layers.0/attn/GroupQueryAttention' Status Message: attention_bias
is not supported in GroupQueryAttention cuda kernel.
```

The rejection is explicit in the kernel and still present on `main`
([group_query_attention.cc#L275-L277](https://github.com/microsoft/onnxruntime/blob/d7efadab308ed9bde763d8b014e30c2e5664b614/onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc#L275-L277)).

**EP support matrix for GQA `attention_bias`:**

| EP | Status | Reference |
|---|---|---|
| CPU | ✅ since #23944 (2025-03) | fp32/fp16, MLAS eltwise-add |
| WebGPU (native) | ✅ since #25285 (2025-07), broadcast hardened in #26769 / #26770 (2025-12) | passed into the flash-attention path |
| **CUDA** | ❌ runtime rejection | this request |

Notably, `position_ids` — added to the spec in the same PR as `attention_bias` (#23944) — already gained CUDA support in the GQA CUDA refactor (#26920, 2026-01). `attention_bias` is the last remaining gap from that input pair.

### Minimal reproduction

https://github.com/jiangzhuo/ort-gqa-attention-bias-cuda-repro

A single-node ONNX graph (GQA + tiny fp16 additive mask), no model download needed:

```bash
pip install -r requirements.txt   # onnx, numpy, onnxruntime-gpu>=1.22
python repro.py
```

Output on onnxruntime-gpu 1.23.2 (RTX 4070 SUPER, driver 580.167.08, Linux x64):

```
=== CPUExecutionProvider ===
OK: output shape (1, 4, 128), dtype float16

=== CUDAExecutionProvider ===
FAILED: InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code
returned while running GroupQueryAttention node. Name:'' Status Message: position_ids and
attention_bias are not supported in GroupQueryAttention cuda kernel.
```

Also reproduced through onnxruntime-node 1.26.0 (CUDA EP) with the real models below.

### Describe scenario use case

Exporters emit `attention_bias` whenever a model's attention pattern is not plain causal (streaming masks, padding + cross-attention combinations). Both of these Transformers.js-exported speech models run fine on the WebGPU EP and the CPU EP but **cannot run on the CUDA EP at all** — the EP you'd most want for 2–4B models:

- [`onnx-community/Voxtral-Mini-4B-Realtime-2602-ONNX`](https://huggingface.co/onnx-community/Voxtral-Mini-4B-Realtime-2602-ONNX) — streaming ASR, decoder GQA carries a streaming attention mask
- [`onnx-community/cohere-transcribe-03-2026-ONNX`](https://huggingface.co/onnx-community/cohere-transcribe-03-2026-ONNX) — ASR, identical rejection on layer 0's GQA node

For desktop applications that ship one ONNX model and pick an EP per user hardware, this makes CUDA the only tier that fails outright.

The semantics are already defined and implemented twice (CPU EP `contrib_ops/cpu/bert`, WebGPU EP `contrib_ops/webgpu/bert`), so this is a kernel-coverage gap rather than a spec question. Even a non-flash fallback (as the standard ONNX `Attention`-op CUDA kernel does when `attn_mask` is present — it falls back to the memory-efficient/unfused path) would unblock these models; today the node fails outright instead of falling back.

### PR DESCRIPTION
**Pull request by jiangzhuo**
### Description

`com.microsoft.GroupQueryAttention`'s optional `attention_bias` input (input #10) is implemented by the CPU EP (#23944) and the WebGPU EP (#25285, #26769), but the CUDA EP rejects it at runtime. This PR wires it through.

No new kernel is needed: the unfused GQA fallback added for #28195 calls `LaunchUnfusedAttention`, whose kernel already implements an additive bias with dim-0/dim-1 broadcast, per-batch `seqlens_k`, causal/sliding-window masking and softcap — the op just passed `attn_bias=nullptr`. The change is dispatch plumbing:

- **`group_query_attention.cc`** — remove the blanket rejection; validate the bias element type; set `broadcast_attn_bias_dim_0/1` from the bias shape (the fields already exist on `AttentionParameters`); add `!has_attention_bias` to the XQA / cuDNN SDPA / flash / flash-fast-decode / MEA eligibility so bias-carrying nodes dispatch to the unfused fallback; set `data.attention_bias`.
- **`attention_data.h`** — add the `attention_bias` pointer to `GroupQueryAttentionData`.
- **`group_query_attention_impl.cu`** — pass the real pointer and broadcast flags in `UnfusedGqaAttention` (previously hardcoded `nullptr`/`false`).

Why each fused path stays disqualified with a bias:
- flash / flash fast-decode: `flash_api.h` has no bias parameter (same exclusion as MHA and the ONNX `Attention`-op CUDA kernel).
- XQA: no bias parameter.
- cuDNN SDPA: GQA's cuDNN path is bottom-right causal, which cuDNN documents as incompatible with a bias (`multihead_attention.cc` has the same restriction).
- cutlass MEA: the wrapper computes the bias row stride from `kv_sequence_length`, which GQA sets to the KV-cache capacity (`seqlen_present_kv_cache`) rather than `total_sequence_length`, so rows would be misaligned under past/present buffer sharing. Left for a follow-up (needs an explicit `attn_bias_strideM` in `MemoryEfficientAttentionParams`).

Kept `NOT_IMPLEMENTED` (explicit, clear errors instead of the previous blanket rejection): bias × quantized KV cache (unfused requires `T == U`), bias × smooth-softmax/head_sink.

### Tests

New `TestGQAAttentionBias` in `test_gqa.py`: prompt and past/decode parity across packed/unpacked QKV, shared/separate KV buffer, rotary, odd head sizes (40/80), and a subsequent multi-token prompt. The bias is **non-zero random** so a kernel that silently ignores the input fails parity (the harness previously modeled a zeros bias). Also fixes the test graph builder declaring the bias input's last dim as the KV-cache capacity instead of `total_sequence_length` (the shape the op validates).

Full `test_gqa.py` suite: 476 tests pass, no regressions (SM89, CUDA 12.8).

### Motivation and Context

Fixes #29506.

Transformers.js-exported speech models carry non-causal attention patterns as `attention_bias` and currently cannot run on the CUDA EP at all, while running fine on WebGPU and CPU:
- [`onnx-community/Voxtral-Mini-4B-Realtime-2602-ONNX`](https://huggingface.co/onnx-community/Voxtral-Mini-4B-Realtime-2602-ONNX) (streaming ASR)
- [`onnx-community/cohere-transcribe-03-2026-ONNX`](https://huggingface.co/onnx-community/cohere-transcribe-03-2026-ONNX)

End-to-end validation with this patch on an RTX 4070 SUPER: the full Voxtral-Mini-4B-Realtime streaming pipeline (q4f16) transcribes correctly at **RTF 0.23** (31 s clip, 25.6 tok/s sustained decode, 6.4 GB VRAM) — on this workload the unfused-attention path outperforms the same model on the WebGPU EP (RTF 0.26).

(While validating, an unrelated pre-existing issue surfaced: `GroupQueryAttentionFusion` breaks graphs whose GQA nodes carry >9 inputs — filed as #29524.)

### CODE REVIEW

File: onnxruntime/test/python/transformers/test_gqa.py [hunk 1]
```diff
@@ -444,10 +444,11 @@ def create_gqa_node_and_io(
...
         )
     if config.has_attention_bias:
-        bias_len = present_kv_seqlen if is_past else config.kv_sequence_length
+        # Per the op spec the last dim is total_sequence_length (past + new), which the kernel
+        # validates at runtime; declare it symbolic so one graph serves all cache states.
         graph_input.append(
             helper.make_tensor_value_info(
-                "attention_bias", ort_type, [config.batch_size, 1, config.q_sequence_length, bias_len]
+                "attention_bias", ort_type, [config.batch_size, 1, config.q_sequence_length, "total_seq_len"]
             )
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply onnxruntime/test/python/transformers/test_gqa.py:458]: Fixed in 47e935c.

File: onnxruntime/test/python/transformers/test_gqa.py [hunk 2]
```diff
@@ -1446,12 +1451,16 @@ def parity_check_gqa_past(
...
     if config.has_attention_bias:
-        attention_bias = torch.zeros(
-            config.batch_size, 1, config.q_sequence_length, total_seq_len, device=device, dtype=torch_type
+        # Random (non-zero) bias so that a kernel silently ignoring the input fails parity.
+        # Positions beyond each batch's valid length get a large negative finite value
+        # (matching the CPU test convention; avoids inf-inf NaN when composed with masking).
+        attention_bias = (
+            torch.randn(config.batch_size, 1, config.q_sequence_length, total_seq_len, device=device, dtype=torch_type)
+            * 0.5
         )
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply onnxruntime/test/python/transformers/test_gqa.py:1476]: Fixed in 47e935c.

File: onnxruntime/test/python/transformers/test_gqa.py [hunk 3]
```diff
@@ -2027,6 +2043,66 @@ def gqa_cuda_past_test_cases(
                     yield name, config
 
 
+def gqa_cuda_attention_bias_test_cases(is_past: bool):
```
Comment (tianleiwu) [inline onnxruntime/test/python/transformers/test_gqa.py:2056]: Test-coverage gap on the broadcast flags. Every case here builds the bias as `[batch_size, 1, S_q, total_seq]`, so `broadcast_attn_bias_dim_0 == false` and `broadcast_attn_bias_dim_1 == true` for all of them. Since wiring `broadcast_attn_bias_dim_0/1` from the bias shape is the core of this PR, the two untested combinations are exactly the ones most likely to hide a bug:

- per-head bias: `dim1 == num_heads` (`broadcast_dim_1 == false`)
- batch-broadcast bias: `dim0 == 1` (`broadcast_dim_0 == true`)

Consider adding at least one case for each so a regression in the non-broadcast head indexing or the dim-0 broadcast flag would be caught.

**Review by tianleiwu — COMMENTED** [review]
Reviewed the dispatch plumbing that routes `attention_bias`-carrying GQA nodes to the unfused fallback. The approach is sound and the key correctness claims hold up against the code:

- All four fused paths correctly gate on `!has_attention_bias` (XQA, cuDNN SDPA, flash — and flash-fast-decode transitively via `use_flash_attention`, plus MEA), so a bias-carrying node deterministically reaches `UnfusedGqaAttention`.
- The unfused softmax indexes the bias row with stride `total_kv_length` (`== parameters.total_sequence_length`), which matches the bias tensor's last dim, and the per-batch `seqlens_k` cutoff keeps reads within `[0, total_sequence_length)`. So the past/present buffer-sharing hazard that disqualifies MEA (row stride derived from the KV-cache capacity) does not apply on this path — the MEA exclusion rationale is accurate.
- `masked_bias_value` stays at its default 0 for GQA (no composed mask sentinel), so the fully-masked-row guard degrades to the intended zero-row behavior via the `s_max == -inf` branch.

One test-coverage gap is worth addressing before merge (inline comment). I also concur with the two review notes already open on the current head, which still apply:
- `TestGQAAttentionBias` is missing a CUDA skip guard; sibling CUDA classes such as `TestGQAQKNorm` all carry `@unittest.skipIf(not has_cuda_device(80), ...)`. Without it, CPU-only / no-GPU runs will fail allocating CUDA tensors.
- No `NumDimensions() == 4` guard exists before the 4-D `attention_bias` shape is indexed (`TensorShape::operator[]` is unchecked, so a malformed rank-<4 bias is UB rather than a clean `INVALID_ARGUMENT`). This is a pre-existing gap in the shared `CheckCustomAttentionInputs` helper (also reached by the CPU/WebGPU EPs), but this PR is the first to exercise it on CUDA.

File: onnxruntime/test/python/transformers/test_gqa.py (hunk 3 above)
Comment (jiangzhuo) [reply onnxruntime/test/python/transformers/test_gqa.py:2056]: Added in 254da44. `GQAConfig` gains `attention_bias_broadcast_dim_0` / `attention_bias_per_head`, threaded through the graph builder and both parity constructions. Both combinations are now covered in both suites: batch-broadcast biases (dim0 == 1) in prompt and past (shared and separate KV buffer), and per-head biases (dim1 == num_heads) in prompt and past.

One note on the dim0-broadcast past cases: the harness's -10000 tail convention is per-batch (`end_pos` follows each batch's `cache_seqlens`), which a batch-shared bias cannot express — so those cases carry a fully random bias with no tail. They pass parity as-is, which doubles as verification that the kernel's `seqlens_k` cutoff and the reference mask never read past each batch's valid length. Full `test_gqa.py` suite: 481 passing on SM 8.9.

File: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc [hunk 4]
```diff
@@ -311,6 +325,12 @@ Status GroupQueryAttention<T, U>::ComputeInternal(OpKernelContext* context) cons
                                                                                head_sink,
                                                                                parameters));
 
+  if (has_attention_bias) {
+    const auto& bias_dims = attention_bias->Shape().GetDims();
+    parameters.broadcast_attn_bias_dim_0 = bias_dims[0] == 1;
+    parameters.broadcast_attn_bias_dim_1 = bias_dims[1] == 1;
+  }
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:332]: Added in 254da44 — a `NumDimensions() == 4` guard at the top of the bias branch in `CheckCustomAttentionInputs`, so the CPU and WebGPU EPs that share the helper get the same clean `INVALID_ARGUMENT` instead of UB.

File: onnxruntime/test/python/transformers/test_gqa.py [hunk 5]
```diff
@@ -2548,6 +2624,45 @@ def test_gqa_past_memory_efficient_bf16(self, name, config):
             )
 
 
+class TestGQAAttentionBias(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (jiangzhuo) [reply onnxruntime/test/python/transformers/test_gqa.py:2653]: Added in 254da44 — `@unittest.skipIf(not has_cuda_device(53), ...)`, matching `TestMemoryEfficientGQA` (the unfused path has no arch floor beyond what the CUDA EP supports). Verified both directions locally: 14 passed with a GPU, 14 skipped with `CUDA_VISIBLE_DEVICES=""` — which is what the CPU-torch CI agents will now do.

**Review by tianleiwu — DISMISSED** [review]

**Comment by skottmckay** [comment]
Can the C++ unit tests be updated to also validate the new code? AFAIK the python tests are only run post-merge, so a change that broke would not be detected in a PR. 

onnxruntime\test\contrib_ops\group_query_attention_op_test.cc

**Comment by jiangzhuo** [comment]
@tianleiwu the earlier review feedback (broadcast-flag coverage, the `TestGQAAttentionBias` CUDA skip guard, and the 4D rank guard in `CheckCustomAttentionInputs`) plus @skottmckay's request for PR-time C++ coverage are all addressed now. Could you approve-and-run the workflows on the latest commit when you get a chance? Fork PRs stay gated until a maintainer runs them.

Added in b3d4a79 — `CudaAttentionBiasParityVsCpu` in `group_query_attention_op_test.cc`. It runs the same prompt case with a non-null `attention_bias` on CUDA (fp16) and CPU (fp32, the reference EP that already implements the input) and compares the outputs, covering the three shapes that drive `broadcast_attn_bias_dim_0/1`: `[batch, 1, S, S]` (default), `[1, 1, S, S]` (dim0 broadcast) and `[batch, num_heads, S, S]` (per-head). It skips cleanly when no CUDA EP is available, so it runs in the GPU CI on every PR rather than only post-merge.

Verified locally on SM 8.9: passes at fp16 tolerance 0.02, and the full `GroupQueryAttentionTest` suite is 56 passed / 12 skipped (WebGPU) / 0 failed.

**Review by tianleiwu — APPROVED** [review]
